### PR TITLE
Update sporadically failing Hermitian test

### PR DESCRIPTION
**Pull request by lillian542**
Adds a line setting `qml.Hermitian._eigs = {}`. If the intermittent failures are because sometimes something is already stored in `_eigs` at the beginning of the test, this should fix it.

Added mocking to confirm that the `eigendecomposition` function is called if the relevant results aren't already cached, but the cached results are used instead if present.

### CODE REVIEW

File: tests/ops/qubit/test_observables.py [hunk 1]
```diff
@@ -272,17 +272,31 @@ def test_hermitian_eigvals_eigvecs_same_observable_twice(
         assert len(qml.Hermitian._eigs) == 1
 
     @pytest.mark.parametrize("observable, eigvals, eigvecs", EIGVALS_TEST_DATA)
-    def test_hermitian_diagonalizing_gates(self, observable, eigvals, eigvecs, tol):
+    def test_hermitian_diagonalizing_gates(self, observable, eigvals, eigvecs, tol, mocker):
         """Tests that the diagonalizing_gates method of the Hermitian class returns the correct results."""
+
+        # ensure _eigs isn't populated from previoius tests
```
Comment (lillian542) [inline tests/ops/qubit/test_observables.py:278]:
```suggestion
        # ensure _eigs isn't populated from previous tests
```

File: tests/ops/qubit/test_observables.py [hunk 2]
```diff
@@ -272,17 +272,31 @@ def test_hermitian_eigvals_eigvecs_same_observable_twice(
...
+        qml.Hermitian._eigs = {}
+
+        # calling diagonalizing gates when observable is not in _eigs adds expected entry to _eigs
```
Comment (lillian542) [inline tests/ops/qubit/test_observables.py:281]:
```suggestion
        # check calling `diagonalizing_gates` when `observable` is not in `_eigs` adds expected entry to `_eigs`
```

**Review by timmysilv — APPROVED** [review]
🎉 🎉 🎉

File: tests/ops/qubit/test_observables.py [hunk 3]
```diff
@@ -272,17 +272,31 @@ def test_hermitian_eigvals_eigvecs_same_observable_twice(
         assert len(qml.Hermitian._eigs) == 1
 
     @pytest.mark.parametrize("observable, eigvals, eigvecs", EIGVALS_TEST_DATA)
-    def test_hermitian_diagonalizing_gates(self, observable, eigvals, eigvecs, tol):
+    def test_hermitian_diagonalizing_gates(self, observable, eigvals, eigvecs, tol, mocker):
         """Tests that the diagonalizing_gates method of the Hermitian class returns the correct results."""
```
Comment (eddddddy) [inline tests/ops/qubit/test_observables.py:276]: The test failure I saw was in a different test. The CI logs are [here](https://github.com/PennyLaneAI/pennylane/actions/runs/5180782930/jobs/9335420469). I'm pretty sure there are a few others.

We can either introduce the same change to the rest of the failing tests, or create a new pytest fixture where we clean up `Hermitian._eigs` before/after every test.

Comment (timmysilv) [reply tests/ops/qubit/test_observables.py:276]: nice catch. if they're all scoped to this file, then it should be pretty straightforward. Just add something like this at the top of the file. don't even need to do stuff after! I'm doing it in another PR [here](https://github.com/PennyLaneAI/pennylane/pull/4257/files#diff-7388684292228c32fbc2732b7a087d9de8d7596d6c5b0836ad76c9fc5e04fc6aR28-R32). So:
```python
@pytest.fixture(autouse=True)
def run_before_tests():
    qml.Hermitian._eigs = {}
    yield
```

Comment (lillian542) [reply tests/ops/qubit/test_observables.py:276]: Thanks! I'll add that :)

File: tests/ops/qubit/test_observables.py [hunk 4]
```diff
@@ -272,17 +279,31 @@ def test_hermitian_eigvals_eigvecs_same_observable_twice(
...
+
+        # ensure _eigs isn't populated from previous tests
+        qml.Hermitian._eigs = {}
+
```
Comment (timmysilv) [inline tests/ops/qubit/test_observables.py:287]: not needed anymore with the fixture :D
```suggestion
```

**Review by eddddddy — APPROVED** [review]
Looks good 🎉

**Pull request merged**